### DISCUSSION TITLE
How to handle typescript generics in mongoose

### DISCUSSION THREAD

**Discussion by ShakedHad** [Automattic/mongoose#12745] [Q&A]
Hi
I'm trying to implement something like this
```
type IData<T> = {
  id: number,
  name: string,
  data: T
}

const DataSchema = new Schema<IData<T>>({
  id: Number,
  name: String,
  data: T
});

const Data = mongoose.model('Data', DataSchema);
```
Basically, a schema that one of its fields is generic.
I'm aware that I can use `Schema.Types.Mixed` for the "data" property, but is there any way to keep the typescript generic feature? Or any better way to implement it?

**Comment by vkarpov15** [comment]
Hard to give an exact answer without more info, but I think the general idea is that you'll need a function that takes a generic parameter and returns a model. Something like the following.

```ts
import mongoose from 'mongoose';

type IData<T> = {
  id: number,
  name: string,
  data: T
}

function Foo<T>(type: Function) {
 const DataSchema = new mongoose.Schema<IData<T>>({
    id: Number,
    name: String,
    data: type
  });
  return mongoose.model('Data', DataSchema);
}

const Test = Foo<String>(String);

const doc = new Test({ data: 42 });
console.log(doc);
```

Does that help?

**Reply by ShakedHad** [reply]
That's an interesting approach, i'll try it

thanks!
